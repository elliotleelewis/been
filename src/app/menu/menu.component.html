<div class="flex items-center justify-center bg-primary text-white">
	<h1 class="p-3 text-xl font-bold tracking-wide">been</h1>
</div>
<form>
	<label for="search" class="sr-only">Search</label>
	<input
		id="search"
		class="w-full border-none px-4 py-2 focus:ring-2 focus:ring-inset focus:ring-primary/50"
		type="text"
		placeholder="Search..."
		[formControl]="searchControl"
	/>
</form>
<ul class="h-full overflow-auto">
	@for (region of filteredRegions$ | async; track region.name) {
		<li>
			<h2 class="bg-slate-200 p-4 font-medium">
				{{ region.name || 'Other' }}
			</h2>
			<ul class="divide-y divide-slate-100">
				@for (country of region.values; track country.iso3166) {
					<li class="hover:bg-slate-50">
						<label class="flex items-center px-4 py-1">
							<input
								class="my-0 me-2 size-4 rounded border-slate-400 bg-slate-50 text-primary focus:ring-2 focus:ring-primary/50 active:ring-primary"
								type="checkbox"
								[ngModel]="country.selected"
								(ngModelChange)="toggleCountry(country)"
							/>
							<span>{{ country.name }}</span>
						</label>
					</li>
				}
			</ul>
		</li>
	}
</ul>
