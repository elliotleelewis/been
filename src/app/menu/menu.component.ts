import {
	ChangeDetectionStrategy,
	Component,
	HostBinding,
	inject,
} from '@angular/core';
import { FormControl } from '@angular/forms';
import {
	type Observable,
	concat,
	defer,
	distinctUntilChanged,
	map,
	of,
	switchMap,
} from 'rxjs';

import { type Country } from '../models/country';
import { type Region } from '../models/region';
import { CountriesService } from '../services/countries.service';

@Component({
	selector: 'app-menu',
	templateUrl: './menu.component.html',
	changeDetection: ChangeDetectionStrategy.OnPush,
})
export class MenuComponent {
	private readonly countriesService = inject(CountriesService);

	@HostBinding('class')
	class = 'flex flex-col overflow-auto dark:bg-zinc-900 dark:text-white';

	searchControl = new FormControl('', { nonNullable: true });

	get filteredRegions$(): Observable<readonly Region[]> {
		return concat(
			defer(() => of(this.searchControl.value)),
			this.searchControl.valueChanges,
		).pipe(
			distinctUntilChanged(),
			map((search) => search.trim().toLowerCase()),
			switchMap((search) =>
				this.countriesService.regions$.pipe(
					map((regions) =>
						search
							? regions
									.map((region) => ({
										...region,
										values: region.values.filter(
											({ name }) =>
												name
													.toLowerCase()
													.includes(search),
										),
									}))
									.filter(
										(region) => region.values.length > 0,
									)
							: regions,
					),
				),
			),
		);
	}

	toggleCountry(country: Country): void {
		if (country.selected) {
			this.countriesService.removeCountry(country.iso3166);
		} else {
			this.countriesService.addCountry(country.iso3166);
		}
	}
}
