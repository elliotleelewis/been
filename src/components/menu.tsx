import { type FC, type ReactNode, memo, useMemo, useState } from 'react';
import { useCountries } from '../contexts/countries-context';
import { MenuItem } from './menu-item';
import { Progress } from './progress';

interface Props {
	header: ReactNode;
}

export const Menu: FC<Props> = memo(({ header }) => {
	const { regions } = useCountries();

	const [search, setSearch] = useState('');

	const filteredRegions = useMemo(() => {
		const s = search.trim().toLowerCase();
		if (!s) {
			return regions;
		}
		return regions
			.map((region) => ({
				...region,
				values: region.values.filter(({ name }) =>
					name.toLowerCase().includes(s),
				),
			}))
			.filter((region) => region.values.length > 0);
	}, [search, regions]);

	return (
		<>
			{header}
			<form className="border-t-2 border-zinc-200 sm:border-t-0 dark:border-zinc-700">
				<label htmlFor="search" className="sr-only">
					Search
				</label>
				<input
					id="search"
					className="w-full border-none bg-white px-4 py-2 text-zinc-900 focus:ring-2 focus:ring-inset focus:ring-primary/50 dark:bg-zinc-900 dark:text-white"
					type="text"
					placeholder="Search..."
					value={search}
					onChange={(e) => {
						setSearch(e.target.value);
					}}
				/>
			</form>
			{filteredRegions.length === 0 ? (
				<div className="m-4 h-full text-center font-medium">
					No results!
				</div>
			) : (
				<ul className="h-full overflow-y-auto">
					{filteredRegions.map((region) => (
						<li key={region.name}>
							<div className="flex items-center justify-between sticky top-0 bg-zinc-200 p-4 font-medium dark:bg-zinc-800">
								<h2>{region.name || 'Other'}</h2>
								<Progress complete={region.complete ?? 0} />
							</div>
							<ul className="divide-y divide-zinc-100 dark:divide-zinc-800">
								{region.values.map((country) => (
									<MenuItem
										key={country.iso3166}
										country={country}
									/>
								))}
							</ul>
						</li>
					))}
				</ul>
			)}
		</>
	);
});
Menu.displayName = 'Menu';
