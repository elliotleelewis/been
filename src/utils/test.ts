import type { WritableAtom } from 'jotai/index';
import { useHydrateAtoms } from 'jotai/utils';
import type { PropsWithChildren } from 'react';

// biome-ignore lint/suspicious/noExplicitAny: Any is required to allow any atom type. Unfortunately cannot be `unknown`.
type AnyWritableAtom = WritableAtom<unknown, any[], unknown>;

type InferAtomTuples<T> = {
	[K in keyof T]: T[K] extends readonly [infer A, ...infer Rest]
		? A extends WritableAtom<unknown, infer Args extends unknown[], unknown>
			? Rest extends Args
				? readonly [A, ...Rest]
				: never
			: never
		: never;
};

export const HydrateAtoms = <
	T extends readonly (readonly [AnyWritableAtom, ...unknown[]])[],
>({
	initialValues,
	children,
}: PropsWithChildren<{
	initialValues: InferAtomTuples<T>;
}>) => {
	useHydrateAtoms(initialValues);
	return children;
};
