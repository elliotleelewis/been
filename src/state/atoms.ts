import { atom } from "jotai";
import { atomWithStorage } from "jotai/utils";
import type { Country } from "../models/country.ts";
import type { Region } from "../models/region.ts";
import { regionalizer } from "../utils/regionalizer.ts";

const COUNTRIES_STORAGE_KEY = "APP_COUNTRIES";

export const rawCountriesAtom = atom<Record<string, Country>>({});
export const selectedCountriesAtom = atomWithStorage<readonly string[]>(
	COUNTRIES_STORAGE_KEY,
	[],
	undefined,
	{ getOnInit: true },
);
export const focusAtom = atom<Country | null>();

export const countriesAtom = atom<readonly Country[]>((get) => {
	const rawCountries = get(rawCountriesAtom);
	const selectedCountries = get(selectedCountriesAtom);

	return Object.values(rawCountries).map((c) =>
		Object.assign(c, {
			selected: selectedCountries.includes(c.iso3166),
		}),
	);
});
export const regionsAtom = atom<readonly Region[]>((get) => {
	const countries = get(countriesAtom);
	return regionalizer(countries);
});

export const addCountryAtom = atom(
	undefined,
	(get, set, countryCode: string) => {
		const selectedCountries = get(selectedCountriesAtom);
		if (!selectedCountries.includes(countryCode)) {
			const countries = get(countriesAtom);
			const focusCountry =
				countries.find((c) => c.iso3166 === countryCode) ?? undefined;
			set(focusAtom, focusCountry);
			set(selectedCountriesAtom, [...selectedCountries, countryCode]);
		}
	},
);
export const removeCountryAtom = atom(
	undefined,
	(get, set, countryCode: string) => {
		const selectedCountries = get(selectedCountriesAtom);
		set(focusAtom, undefined);
		set(
			selectedCountriesAtom,
			selectedCountries.filter((c) => c !== countryCode),
		);
	},
);
