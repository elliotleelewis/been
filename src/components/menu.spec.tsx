import { render } from '@testing-library/react';
import { Provider } from 'jotai';
import { describe, expect, it } from 'vitest';
import type { Country } from '../models/country.ts';
import { rawCountriesAtom } from '../state/atoms.ts';
import { HydrateAtoms } from '../utils/test.ts';
import { Menu } from './menu';

const country: Country = {
	name: 'United Kingdom',
	iso3166: 'GB',
	region: 'Europe',
};

describe('Menu', () => {
	it('should render', () => {
		const result = render(
			<Provider>
				<HydrateAtoms
					initialValues={[
						[rawCountriesAtom, { [country.iso3166]: country }],
					]}
				>
					<Menu
						fullscreen={false}
						toggleFullscreen={() => undefined}
					/>
				</HydrateAtoms>
			</Provider>,
		);

		expect(result.asFragment()).toMatchSnapshot();
	});
});
