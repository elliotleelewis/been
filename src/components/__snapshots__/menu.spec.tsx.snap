// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Menu > should render 1`] = `
<DocumentFragment>
  <form
    class="border-zinc-200 border-t-2 sm:border-t-0 dark:border-zinc-700"
  >
    <label
      class="sr-only"
      for="search"
    >
      Search
    </label>
    <input
      class="w-full border-none bg-white px-4 py-2 text-zinc-900 focus:ring-2 focus:ring-primary/50 focus:ring-inset disabled:hover:cursor-not-allowed dark:bg-zinc-900 dark:text-white"
      id="search"
      placeholder="Search..."
      type="text"
      value=""
    />
  </form>
  <ul
    class="h-full overflow-y-auto"
  >
    <li>
      <div
        class="sticky top-0 flex items-center justify-between bg-zinc-200 p-4 font-medium dark:bg-zinc-800"
      >
        <h2>
          Europe
        </h2>
        <div
          class="relative size-5"
        >
          <svg
            class="-rotate-90 size-full"
            viewBox="0 0 36 36"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              0% complete.
            </title>
            <circle
              class="stroke-current text-gray-500 dark:text-neutral-500"
              cx="18"
              cy="18"
              fill="none"
              r="16"
              stroke-width="4"
            />
            <circle
              class="stroke-current text-primary dark:text-primary"
              cx="18"
              cy="18"
              fill="none"
              r="16"
              stroke-dasharray="100"
              stroke-dashoffset="100"
              stroke-linecap="round"
              stroke-width="4"
            />
          </svg>
        </div>
      </div>
      <ul
        class="divide-y divide-zinc-100 dark:divide-zinc-800"
      >
        <li
          class="hover:bg-zinc-50 dark:hover:bg-zinc-800"
        >
          <label
            class="flex items-center px-4 py-1"
          >
            <input
              class="my-0 me-2 size-4 rounded border-zinc-400 bg-zinc-50 text-primary focus:ring-2 focus:ring-primary/50 active:ring-primary dark:border-zinc-600 dark:bg-zinc-900"
              id="GB"
              type="checkbox"
            />
            <span>
              United Kingdom
            </span>
          </label>
        </li>
      </ul>
    </li>
  </ul>
</DocumentFragment>
`;
